### PR TITLE
Make Quill instances detachable and reconfigurable

### PR DESCRIPTION
**Pull request by singintime**
## Motivation

This PR addresses discussion #4397 and issue #4234, providing a way to apply a new configuration to an existing Quill editor. Please refer to those for further information about the reasons behind this change.

## Proposed changes

- Adding a `Subscriber` class, acting as a proxy to the `EventTarget` API and keeping track of all subscriptions to event listeners originated by Quill components and modules,
- Adding getters and setters to bind `Subscriber` instances to Quill root elements. `Subscriber` instances are stored in a `WeakMap`, similar to what happens for Quill instances.
- Using the subscriber to replace all non-static calls of `addEventListener()`,
- Unsubscribing from all event listeners originated by blots when they `detach()`,
- Unsubscribing from all events originated by a Quill instance when detached or reconfigured,
- Adding a `detach()` cleanup method to Quill instances,
- Adding a `configure()` method to apply a new configuration to an already existing instance of Quill,
- Providing unit test coverage for the above-mentioned changes.

### CODE REVIEW

File: packages/quill/src/core/subscriber.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+/**
+ * Gets the Subscriber instance bound to the given object.
+ * Throws an error if the binding does not exist.
+ */
+export function getSubscriber(object: Source): Subscriber {
+  const subscriber = subscribers.get(object);
+  if (!subscriber) {
+    throw new Error(
```
Comment (bmakuh) [inline packages/quill/src/core/subscriber.ts:39]: Should we throw here, or should we just return undefined? The potential to throw means any call of this function would need to be wrapped in a try-catch which doesn't seem ideal

File: packages/quill/src/core/quill.ts [hunk 2]
```diff
@@ -193,22 +194,45 @@ class Quill {
...
   constructor(container: HTMLElement | string, options: QuillOptions = {}) {
-    this.options = expandConfig(container, options);
-    this.container = this.options.container;
-    if (this.container == null) {
-      debug.error('Invalid Quill container', container);
-      return;
+    this.setup(container, options);
+  }
+
+  configure(options: QuillOptions = {}) {
```
Comment (bmakuh) [inline packages/quill/src/core/quill.ts:200]: Since this is a public method, maybe we could add a JSDoc for it as well as mention it in the documentation?

File: packages/quill/src/core/quill.ts [hunk 3]
```diff
@@ -193,22 +194,45 @@ class Quill {
...
     }
-    if (this.options.debug) {
-      Quill.debug(this.options.debug);
+
+    const subscriber = getSubscriber(this.root);
+    subscriber.removeAllListeners();
+    this.emitter.removeAllListeners();
+    this.emitter.ignoreDOM();
+    this.scroll.detach();
+    return true;
```
Comment (bmakuh) [inline packages/quill/src/core/quill.ts:214]: Correct me if I'm wrong but this method does not appear to fully tear down passed in options, modules, etc. It seems to me that this method should return the quill instance to "factory settings" no?

Additionally, any dynamically changed things (like calling `quill.keyboard.addBinding`) should also be cleaned up

Comment (singintime) [reply packages/quill/src/core/quill.ts:214]: This method is not supposed to restore the quill instance to "factory settings", that would be the job of `configure()` called without arguments. `detach()` makes the Quill instance unresponsive to all events and user actions.

This is also why I didn't reset dynamically changed modules (keyboard, clipboard, history, etc.): a detached Quill instance is "dead", and not meant to be used anymore until reconfigured via the `configure()` method

File: packages/quill/src/core/subscriber.ts (hunk 1 above)
Comment (singintime) [reply packages/quill/src/core/subscriber.ts:39]: I agree, this is not ideal... I also don't like the undefined option though, as we would only trade try/catch blocks for non-null checks. I've got another idea: getting rid of `createSubscriber()` entirely, and creating the new subscriber instance on the first call of `getSubscriber()` for a given object. The drawback is that one might inadvertently create multiple subscribers instead of one if they call `getSubscriber()` with the wrong argument. For this reason, I added some logging on the creation of new subscribers to facilitate debugging. Please let me know what you think :)

File: packages/website/content/docs/api.mdx [hunk 4]
```diff
@@ -21,6 +21,28 @@ title: API
 
 ## Content
 
+### configure
+
+Applies a new [configuration](/docs/configuration/) to an existing instance of Quill.
```
Comment (bmakuh) [inline packages/website/content/docs/api.mdx:26]:
```suggestion
Applies a new [configuration](/docs/configuration/) to an existing instance of Quill. To reset a quill instance to its defaults, call `quill.configure()` without any arguments.
```

File: packages/website/content/docs/api.mdx [hunk 5]
```diff
@@ -21,6 +21,28 @@ title: API
...
+
+**Methods**
+
+```typescript
+configure(options: QuillOptions = {}): void
+```
+
+**Example 1**
```
Comment (bmakuh) [inline packages/website/content/docs/api.mdx:34]:
```suggestion
**Example 1: resetting an instance to defaults**
```

File: packages/website/content/docs/api.mdx [hunk 6]
```diff
@@ -21,6 +21,28 @@ title: API
...
+
+```typescript
+quill.configure();
+```
+
+**Example 2**
```
Comment (bmakuh) [inline packages/website/content/docs/api.mdx:40]:
```suggestion
**Example 2: setting up a read-only instance with the built-in Bubble theme**
```

File: packages/quill/src/core/subscriber.ts [hunk 7]
```diff
@@ -20,25 +24,16 @@ interface Subscription {
...
-  subscribers.set(object, subscriber);
-  return subscriber;
-}
-
 /**
  * Gets the Subscriber instance bound to the given object.
- * Throws an error if the binding does not exist.
+ * Creates a new one if the binding does not exist yet.
  */
 export function getSubscriber(object: Source): Subscriber {
```
Comment (bmakuh) [inline packages/quill/src/core/subscriber.ts:31]: Personally I'd be okay with this method returning undefined—the signature of _any_ getter from a map always has the potential to not find what you're looking for: 
```ts
interface WeakMap {
  get(key: object): undefined | any // value can be anything, including "not found"
}
```

At any rate, I'm okay with this direction too, but I think we should name the method slightly differently in this case to better reflect the fact that it's not _always_ just a getter. How about `findOrCreate()`?

Comment (singintime) [reply packages/quill/src/core/subscriber.ts:31]: Renamed to `findOrCreateSubscriber()`

File: packages/quill/src/core/quill.ts [hunk 8]
```diff
@@ -234,6 +235,7 @@ class Quill {
     }
     const scrollBlotName = Parchment.ScrollBlot.blotName;
     const ScrollBlot = this.options.registry.query(scrollBlotName);
+    console.log(ScrollBlot);
```
Comment (bmakuh) [inline packages/quill/src/core/quill.ts:238]: 🗑️ ?

File: packages/quill/src/core/theme.ts [hunk 9]
```diff
@@ -27,6 +28,14 @@ class Theme {
     protected options: ThemeOptions,
   ) {}
 
+  detach() {
+    Object.values(this.modules).forEach((module) => {
+      if (module instanceof Module) {
```
Comment (bmakuh) [inline packages/quill/src/core/theme.ts:33]: I wonder if instead a better check might be `if ('detach' in module)`. You _can_ register things with quill that aren't literally subclasses of `Module`

Comment (singintime) [reply packages/quill/src/core/theme.ts]: Agreed! Done :)
